File: StreamProcessorService/src/main/java/com/stream/processor/service/EventProcessingService.java
package com.stream.processor.service;


import com.stream.processor.Model.SensorEvent;
import com.stream.processor.enrichment.EnrichmentStrategy;
import com.stream.processor.normalization.NormalizationStrategy;
import com.stream.processor.serialization.EventSerializer;
import com.stream.processor.serialization.SerializationException;
import com.stream.processor.validation.ValidationResult;
import com.stream.processor.validation.ValidationRule;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.Optional;

@Slf4j
@Service
@RequiredArgsConstructor
public class EventProcessingService {

    private final EventSerializer serializer;
    private final ValidationRule validator;
    private final NormalizationStrategy normalizer;
    private final EnrichmentStrategy enricher;


    public Optional<String> process(String jsonEvent){
        try {
            SensorEvent event = serializer.deserialize(jsonEvent);

            ValidationResult validationResult = validator.validate(event);
            if (!validationResult.isValid()) {
                log.warn("Validation failed for event {}: {}",
                        event.getMessageId(),
                        String.join(", ", validationResult.getErrors()));
                return Optional.empty();
            }

            SensorEvent processed = enricher.enrich(normalizer.normalize(event));

            String result = serializer.serialize(processed) ;
            log.debug("Successfully processed event: {}", event.getMessageId());
            return Optional.of(result);
        } catch (SerializationException e){
            log.error("Serialization error: {}", e.getMessage(), e);
            return Optional.empty();
        } catch (Exception e) {
            log.error("Processing error for payload: {}", jsonEvent, e);
            return Optional.empty();
        }
    }
}

File: validation_service/event_processing_service.py
"""Main service for processing events"""
import sys
import os
import json
import logging
from typing import Optional

# Add current directory to path for imports
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from composite_validator import CompositeValidator
from field_name_normalization_strategy import FieldNameNormalizationStrategy
from timestamp_normalization_strategy import TimestampNormalizationStrategy
from composite_enrichment_strategy import CompositeEnrichmentStrategy

logger = logging.getLogger(__name__)


class EventProcessingService:
    """Main service for processing events"""
    
    def __init__(self):
        self.validator = CompositeValidator()
        self.field_normalizer = FieldNameNormalizationStrategy()
        self.timestamp_normalizer = TimestampNormalizationStrategy()
        self.enricher = CompositeEnrichmentStrategy()
    
    def process(self, json_event: str) -> Optional[str]:
        try:
            event = json.loads(json_event)
            
            # First normalize field names from CSV format to snake_case
            event = self.field_normalizer.normalize(event)
            
            # Then validate
            validation_result = self.validator.validate(event)
            if not validation_result.valid:
                logger.warning(
                    f"Validation failed for event {event.get('message_id')}: "
                    f"{', '.join(validation_result.errors)}"
                )
                return None
            
            # Normalize timestamp and enrich
            processed = self.enricher.enrich(self.timestamp_normalizer.normalize(event))
            
            result = json.dumps(processed)
            logger.debug(f"Successfully processed event: {processed.get('message_id')} for patient {processed.get('patient_id')}")
            return result
        except json.JSONDecodeError as e:
            logger.error(f"Serialization error: {e}")
            return None
        except Exception as e:
            logger.error(f"Processing error for payload: {json_event}", exc_info=True)
            return None

